Building with clang-16 fails with "error: conflicting types for 'malloc'"

**Description**
<!-- A clear and concise description of what the bug is. -->
Building perl-5.36.0 with clang-16 in Gentoo Linux fails with 
```
sdbm.c:37:17: error: conflicting types for 'malloc'
extern Malloc_t malloc(MEM_SIZE);
                ^
/usr/include/stdlib.h:553:14: note: previous declaration is here
extern void *malloc (size_t __size) __THROW __attribute_malloc__
             ^
sdbm.c:38:15: error: conflicting types for 'free'
extern Free_t free(Malloc_t);
              ^
/usr/include/stdlib.h:568:13: note: previous declaration is here
extern void free (void *__ptr) __THROW;
            ^
2 errors generated.
```

This is because clang-16 sets `-Werror,-Wdeprecated-non-prototype` by default and the offending code snippet in `Configure` has two function declarations without their argument type:

```
#ifdef TRY_MALLOC
void *malloc();
#endif
#ifdef TRY_FREE
void free();
#endif
```
The ensuing internal error becomes:

`malloc.c:12:7: error: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C2x, conflicting with a previous declaration [-Werror,-Wdeprecated-non-prototype]`

And this leads to the Perl build setting `malloctype` to `char *` instead of `void *` and `freetype` to `int` instead of `void`, eventually resulting in function declaration mismatches.

**Steps to Reproduce**
<!-- A one-liner or script to reproduce the issue. -->
`ACCEPT_KEYWORDS=** emerge =sys-devel/clang-common-16.0.0_pre20230107::gentoo`
`ACCEPT_KEYWORDS=~amd64  CC=clang-16 emerge -1 =dev-lang/perl-5.36.0-r1`

or 

`ACCEPT_KEYWORDS=~amd64 USE=stricter emerge =sys-devel/clang-common-15.0.7::gentoo`
`ACCEPT_KEYWORDS=~amd64  CC=clang-15 emerge -1 =dev-lang/perl-5.36.0-r1`

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->
Successful build.

> **Description**
> 
> Building perl-5.36.0 with clang-16 in Gentoo Linux fails with
> 
> ```
> sdbm.c:37:17: error: conflicting types for 'malloc'
> extern Malloc_t malloc(MEM_SIZE);
>                 ^
> /usr/include/stdlib.h:553:14: note: previous declaration is here
> extern void *malloc (size_t __size) __THROW __attribute_malloc__
>              ^
> sdbm.c:38:15: error: conflicting types for 'free'
> extern Free_t free(Malloc_t);
>               ^
> /usr/include/stdlib.h:568:13: note: previous declaration is here
> extern void free (void *__ptr) __THROW;
>             ^
> 2 errors generated.
> ```
> 
> This is because clang-16 sets `-Werror,-Wdeprecated-non-prototype` by default and the offending code snippet in `Configure` has two function declarations without their argument type:
> 
> ```
> #ifdef TRY_MALLOC
> void *malloc();
> #endif
> #ifdef TRY_FREE
> void free();
> #endif
> ```
> 
> The ensuing internal error becomes:
> 
> `malloc.c:12:7: error: a function declaration without a prototype is deprecated in all versions of C and is treated as a zero-parameter prototype in C2x, conflicting with a previous declaration [-Werror,-Wdeprecated-non-prototype]`
> 
> And this leads to the Perl build setting `malloctype` to `char *` instead of `void *` and `freetype` to `int` instead of `void`, eventually resulting in function declaration mismatches.
> 
> **Steps to Reproduce**
> 
> `ACCEPT_KEYWORDS=** emerge =sys-devel/clang-common-16.0.0_pre20230107::gentoo` `ACCEPT_KEYWORDS=~amd64 CC=clang-16 emerge -1 =dev-lang/perl-5.36.0-r1`
> 
> or
> 
> `ACCEPT_KEYWORDS=~amd64 USE=stricter emerge =sys-devel/clang-common-15.0.7::gentoo` `ACCEPT_KEYWORDS=~amd64 CC=clang-15 emerge -1 =dev-lang/perl-5.36.0-r1`
> 
> **Expected behavior**
> 
> Successful build.

Is there anyone on list with clang-16 installed who could confirm this report?

--- a/Configure
+++ b/Configure
@@ -7927,10 +7927,10 @@ $cat >malloc.c <<END
 #include <stdlib.h>
 #endif
 #ifdef TRY_MALLOC
-void *malloc();
+void *malloc(size_t);
 #endif
 #ifdef TRY_FREE
-void free();
+void free(void *);
 #endif
 END
 case "$malloctype" in

